pydantic_ai_slim/pydantic_ai/_agent_graph.py

```diff
@@ -218,21 +218,30 @@ async def _prepare_request_parameters(
     ctx: GraphRunContext[GraphAgentState, GraphAgentDeps[DepsT, NodeRunEndT]],
 ) -> models.ModelRequestParameters:
     """Build tools and create an agent model."""
-    function_tool_defs: list[ToolDefinition] = []
+    function_tool_defs: dict[str, ToolDefinition] = {}
 
     run_context = build_run_context(ctx)
 
     async def add_tool(tool: Tool[DepsT]) -> None:
         ctx = run_context.replace_with(retry=tool.current_retry, tool_name=tool.name)
         if tool_def := await tool.prepare_tool_def(ctx):
-            function_tool_defs.append(tool_def)
+            # prepare_tool_def may change tool_def.name
+            if tool_def.name in function_tool_defs:
+                raise exceptions.UserError(
+                    f"Tool '{tool}' defines a tool whose name conflicts with existing tool: {tool_def.name!r}."
```

Comment:
Can we clarify that the conflict came from the preparation? Something like `New tool name X for tool Y conflicts with existing tool.`

Reply:
@DouweM Looks like `"Tool name 'bar' for tool `Tool(function=<function test_repeat_tool.<locals>.bar at 0x13d631580>, takes_ctx=False, max_retries=1, name='bar', description='', prepare=None, docstring_format='auto', require_parameter_descriptions=False, strict=None, _is_async=False, _single_arg_name=None, _positional_fields=[], _var_positional_field=None, _base_parameters_json_schema={'additionalProperties': False, 'properties': {'x': {'type': 'integer'}, 'y': {'type': 'string'}}, 'required': ['x', 'y'], 'type': 'object'}, current_retry=0)` conflicts with existing tool: 'bar'."`? Or just `Tool name 'bar' for tool(function=<function test_repeat_tool.<locals>.bar at 0x13d631580>)`